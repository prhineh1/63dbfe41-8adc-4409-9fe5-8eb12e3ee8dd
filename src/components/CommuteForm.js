import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import Timepicker from 'rc-time-picker';

export default class CommuteForm extends React.Component {
  state = {
    tempHi: '',
    tempLow: '',
    precipProbability: '',
    commuteDay: moment(),
    commuteTime: moment(),
    calendarFocused: false,
    error: '',
  };
  onLowChange = e => {
    const tempLow = e.target.value;
    if (!tempLow || tempLow.match(/^[1-9][0-9]{0,2}?$/)) { //matches any number between 0 and 1000
      this.setState(() => ({ tempLow }));
    }
  };
  onHighChange = e => {
    const tempHi = e.target.value;
    if (!tempHi || tempHi.match(/^[1-9][0-9]{0,2}?$/)) {
      this.setState(() => ({ tempHi }));
    }
  };
  onPrecipChange = e => {
    const precipProbability = e.target.value;
    if (!precipProbability || precipProbability.match(/(^[1-9][0-9]?$)|100?$|^0?$/)) { //matches any number between 0 and 100
      this.setState(() => ({ precipProbability }))
    }
  }
  onDayChange = commuteDay => {
    if (commuteDay) {
      this.setState(() => ({ commuteDay }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  isOutsideRange = day => !day.isBetween(moment().subtract(1, 'd'), moment().add(14, 'd')); // Can only select dates within two weeks of today
  onTimeChange = commuteTime => {
    if (commuteTime) {
      this.setState(() => ({ commuteTime }));      
    }
  };
  onSubmit = e => {
    e.preventDefault();

    if(!this.state.tempHi || !this.state.tempLow || !this.state.precipProbability) {
      this.setState(() => ({ error: 'Please provide values for the first three fields.' }))
    } else if (parseInt(this.state.tempLow) > parseInt(this.state.tempHi)) {
      this.setState(() => ({ error: "The Hi temperature (second input) must be greater than the low temperature (first input)." }))
    } else {
      this.setState(() => ({ error: '' }));

      // create a momentObj combining the year, month and date of commuteDay
      // with the hour from commuteTime to get the appropriate commuteDate
      const commuteDate = moment({  
        year: this.state.commuteDay.year(),
        month: this.state.commuteDay.month(),
        date: this.state.commuteDay.date(),
        hour: this.state.commuteTime.hour()
      });
      this.props.onSubmit({
        tempLow: parseInt(this.state.tempLow),
        tempHi: parseInt(this.state.tempHi),
        precipProbability: parseFloat(this.state.precipProbability) / 100,
        commuteDate: commuteDate.unix() // convert momentObj to Unix timestamp for use with DarkSky API
      });
    }
  }
  render() {
    return (
      <form onSubmit={this.onSubmit}>
        {this.state.error && <p>{this.state.error}</p>}
        <input 
          type='text'
          placeholder='Low'
          autoFocus
          value={this.state.tempLow}
          onChange={this.onLowChange}
        />
        <input 
          type='text'
          placeholder='Hi'
          value={this.state.tempHi}
          onChange={this.onHighChange}
        />
        <input 
          type='text'
          placeholder='Chance of Rain'
          value={this.state.precipProbability}
          onChange={this.onPrecipChange}
        />
        <SingleDatePicker 
          date={this.state.commuteDay} // momentPropTypes.momentObj or null
          onDateChange={this.onDayChange} // PropTypes.func.isRequired
          focused={this.state.calendarFocused} // PropTypes.bool
          onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
          numberOfMonths={1}
          isOutsideRange={this.isOutsideRange} // PropTypes.func
        />
        <Timepicker
          showSecond={false}
          defaultValue={this.state.commuteTime} // momentPropTypes.momentObj or null
          format={'h:mm a'} // See https://momentjs.com/docs/#/parsing/string-format/
          onChange={this.onTimeChange} 
          use12Hours // PropTypes.bool, followed implemenation at http://react-component.github.io/time-picker/examples/12hours.html
        />
        <div>
          <button>Submit</button>
        </div>
      </form>
    );
  }
}