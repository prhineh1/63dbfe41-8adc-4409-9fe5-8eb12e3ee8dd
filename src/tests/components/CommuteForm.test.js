import React from 'react';
import { shallow, setState } from 'enzyme';
import moment from 'moment';
import CommuteForm from '../../components/CommuteForm';

let wrapper, onSubmit;

beforeEach(() => {
  onSubmit = jest.fn();
  wrapper = shallow(<CommuteForm onSubmit={onSubmit} />);
});

test('should render CommuteForm', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should set tempLow on input change', () => {
  const value = '50';
  wrapper.find('input').at(0).simulate('change', {
    target: { value }
  });
  expect(wrapper.state('tempLow')).toBe(value);
});

test('should set tempHi on input change', () => {
  const value = '85';
  wrapper.find('input').at(1).simulate('change', {
    target: { value }
  });
  expect(wrapper.state('tempHi')).toBe(value);
});

test('should set precipProbabilityCommute on input change', () => {
  const value = '67';
  wrapper.find('input').at(2).simulate('change', {
    target: { value }
  });
  expect(wrapper.state('precipProbabilityCommute')).toBe(value);
});

test('should set new date on date change', () => {
  const time = moment();
  wrapper.find('SingleDatePicker').prop('onDateChange')(time);
  expect(wrapper.state('commuteDay')).toEqual(time);
});

test('should set calendar focus on change', () => {
  const focused = true;
  wrapper.find('SingleDatePicker').prop('onFocusChange')({ focused });
  expect(wrapper.state('calendarFocused')).toBe(focused);
});

test('should set new date on time change', () => {
  const time = moment();
  wrapper.find('Picker').prop('onChange')(time); // Timepicker component is Rendered as 'Picker' (see snapshot)
  expect(wrapper.state('commuteTime')).toEqual(time);
});

test('isOutsideRange should return true', () => {
  const time = moment().add(3, 'w');
  const instance = wrapper.instance().isOutsideRange(time);
  expect(instance).toEqual(true);
});

test('isOutsideRange should return false', () => {
  const time = moment().add(1, 'w');
  const instance = wrapper.instance().isOutsideRange(time);
  expect(instance).toEqual(false);
});

test('should render error for invalid form submission (missing input)', () => {
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submission (tempLow > tempHi)', () => {
  const state = {
    tempHi: '45',
    tempLow: '89',
    precipProbabilityCommute: '25'
  };
  wrapper.setState(state);
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submission (tempLow === "-")', () => {
  const state = {
    tempHi: '45',
    tempLow: '-',
    precipProbabilityCommute: '25'
  };
  wrapper.setState(state);
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test('should call onSubmit prop for valid form submission', () => {
  const state = {
    tempHi: '89',
    tempLow: '43',
    precipProbabilityCommute: '25',
    commuteDay: moment(),
    commuteTime: moment().hour(9),
    calendarFocused: false,
    error: 'error'
  };
  wrapper.setState(state);
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(wrapper.state('error')).toBe('');
  expect(onSubmit).toHaveBeenLastCalledWith({
    tempHi: 89,
    tempLow: 43,
    precipProbabilityCommute: 0.25,
    commuteDate: moment().hour(9).unix()
  });
});
