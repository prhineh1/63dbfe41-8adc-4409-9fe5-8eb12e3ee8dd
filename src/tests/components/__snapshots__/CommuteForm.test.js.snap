// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CommuteForm 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Low (℉)"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Hi (℉)"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Chance of Precipitation (%)"
    type="text"
    value=""
  />
  <SingleDatePicker
    anchorDirection="left"
    customCloseIcon={null}
    customInputIcon={null}
    date={"1970-01-01T00:00:00.000Z"}
    daySize={39}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    firstDayOfWeek={null}
    focused={false}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    id="date"
    initialVisibleMonth={null}
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepOpenOnDateSelect={false}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPrev={null}
    numberOfMonths={1}
    onClose={[Function]}
    onDateChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    orientation="horizontal"
    phrases={
      Object {
        "chooseAvailableDate": [Function],
        "clearDate": "Clear Date",
        "closeDatePicker": "Close",
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month",
        "jumpToPrevMonth": "Move backward to switch to the previous month",
        "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
        "moveFocusByOneMonth": "Switch months",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
        "openThisPanel": "Open this panel",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field",
        "selectFocusedDate": "Select the date in focus",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    placeholder="Date"
    readOnly={false}
    renderCalendarInfo={null}
    renderDay={null}
    renderMonth={null}
    reopenPickerOnClearDate={false}
    required={false}
    screenReaderInputMessage=""
    showClearDate={false}
    showDefaultInputIcon={false}
    withFullScreenPortal={false}
    withPortal={false}
  />
  <Picker
    addon={[Function]}
    align={Object {}}
    allowEmpty={true}
    className="time-picker"
    clearText="clear"
    defaultOpen={false}
    defaultOpenValue={"1970-01-01T00:00:00.000Z"}
    defaultValue={"1970-01-01T00:00:00.000Z"}
    disabledHours={[Function]}
    disabledMinutes={[Function]}
    disabledSeconds={[Function]}
    focusOnOpen={false}
    format="h:mm a"
    hideDisabledOptions={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClose={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onOpen={[Function]}
    placement="bottomLeft"
    popupClassName=""
    prefixCls="rc-time-picker"
    showHour={true}
    showMinute={true}
    showSecond={false}
    style={Object {}}
    use12Hours={true}
  />
  <div>
    <button
      className="button"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`should render error for invalid form submission (missing input) 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__error"
  >
    Please provide values for the first three fields.
  </p>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Low (℉)"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Hi (℉)"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Chance of Precipitation (%)"
    type="text"
    value=""
  />
  <SingleDatePicker
    anchorDirection="left"
    customCloseIcon={null}
    customInputIcon={null}
    date={"1970-01-01T00:00:00.000Z"}
    daySize={39}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    firstDayOfWeek={null}
    focused={false}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    id="date"
    initialVisibleMonth={null}
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepOpenOnDateSelect={false}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPrev={null}
    numberOfMonths={1}
    onClose={[Function]}
    onDateChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    orientation="horizontal"
    phrases={
      Object {
        "chooseAvailableDate": [Function],
        "clearDate": "Clear Date",
        "closeDatePicker": "Close",
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month",
        "jumpToPrevMonth": "Move backward to switch to the previous month",
        "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
        "moveFocusByOneMonth": "Switch months",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
        "openThisPanel": "Open this panel",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field",
        "selectFocusedDate": "Select the date in focus",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    placeholder="Date"
    readOnly={false}
    renderCalendarInfo={null}
    renderDay={null}
    renderMonth={null}
    reopenPickerOnClearDate={false}
    required={false}
    screenReaderInputMessage=""
    showClearDate={false}
    showDefaultInputIcon={false}
    withFullScreenPortal={false}
    withPortal={false}
  />
  <Picker
    addon={[Function]}
    align={Object {}}
    allowEmpty={true}
    className="time-picker"
    clearText="clear"
    defaultOpen={false}
    defaultOpenValue={"1970-01-01T00:00:00.000Z"}
    defaultValue={"1970-01-01T00:00:00.000Z"}
    disabledHours={[Function]}
    disabledMinutes={[Function]}
    disabledSeconds={[Function]}
    focusOnOpen={false}
    format="h:mm a"
    hideDisabledOptions={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClose={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onOpen={[Function]}
    placement="bottomLeft"
    popupClassName=""
    prefixCls="rc-time-picker"
    showHour={true}
    showMinute={true}
    showSecond={false}
    style={Object {}}
    use12Hours={true}
  />
  <div>
    <button
      className="button"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`should render error for invalid form submission (tempLow > tempHi) 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__error"
  >
    The Hi temperature (second input) must be greater than the low temperature (first input).
  </p>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Low (℉)"
    type="text"
    value="89"
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Hi (℉)"
    type="text"
    value="45"
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Chance of Precipitation (%)"
    type="text"
    value="25"
  />
  <SingleDatePicker
    anchorDirection="left"
    customCloseIcon={null}
    customInputIcon={null}
    date={"1970-01-01T00:00:00.000Z"}
    daySize={39}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    firstDayOfWeek={null}
    focused={false}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    id="date"
    initialVisibleMonth={null}
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepOpenOnDateSelect={false}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPrev={null}
    numberOfMonths={1}
    onClose={[Function]}
    onDateChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    orientation="horizontal"
    phrases={
      Object {
        "chooseAvailableDate": [Function],
        "clearDate": "Clear Date",
        "closeDatePicker": "Close",
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month",
        "jumpToPrevMonth": "Move backward to switch to the previous month",
        "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
        "moveFocusByOneMonth": "Switch months",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
        "openThisPanel": "Open this panel",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field",
        "selectFocusedDate": "Select the date in focus",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    placeholder="Date"
    readOnly={false}
    renderCalendarInfo={null}
    renderDay={null}
    renderMonth={null}
    reopenPickerOnClearDate={false}
    required={false}
    screenReaderInputMessage=""
    showClearDate={false}
    showDefaultInputIcon={false}
    withFullScreenPortal={false}
    withPortal={false}
  />
  <Picker
    addon={[Function]}
    align={Object {}}
    allowEmpty={true}
    className="time-picker"
    clearText="clear"
    defaultOpen={false}
    defaultOpenValue={"1970-01-01T00:00:00.000Z"}
    defaultValue={"1970-01-01T00:00:00.000Z"}
    disabledHours={[Function]}
    disabledMinutes={[Function]}
    disabledSeconds={[Function]}
    focusOnOpen={false}
    format="h:mm a"
    hideDisabledOptions={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClose={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onOpen={[Function]}
    placement="bottomLeft"
    popupClassName=""
    prefixCls="rc-time-picker"
    showHour={true}
    showMinute={true}
    showSecond={false}
    style={Object {}}
    use12Hours={true}
  />
  <div>
    <button
      className="button"
    >
      Submit
    </button>
  </div>
</form>
`;
